x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/LocalPhysicalPlanOptimizer.java

```diff
@@ -427,4 +429,49 @@ public static boolean isPushableFieldAttribute(Expression exp, Predicate<FieldAt
         return false;
     }
 
+    private static class SpatialDocValuesExtraction extends OptimizerRule<AggregateExec> {
+        @Override
+        protected PhysicalPlan rule(AggregateExec aggregate) {
+            var foundAttribute = new Holder<Attribute>(null);
```

Comment:
I am not sure if this is to blame or not (a single Holder for attributes, while there may be more than one while searching). So, I tested with `FROM airports2 | stats centroid1=ST_CENTROID(location), centroid2=ST_CENTROID(location2)` where `location` and `location2` are `geo_point`s. This fails with

```
Caused by: java.lang.ClassCastException: class org.elasticsearch.compute.data.BytesRefArrayBlock cannot be cast to class org.elasticsearch.compute.data.LongBlock (org.elasticsearch.compute.data.BytesRefArrayBlock and org.elasticsearch.compute.data.LongBlock are in unnamed module of loader java.net.FactoryURLClassLoader @550fa96f)
        at org.elasticsearch.compute.aggregation.spatial.SpatialCentroidGeoPointDocValuesAggregatorFunction.addRawInput(SpatialCentroidGeoPointDocValuesAggregatorFunction.java:64)
        at org.elasticsearch.compute.aggregation.Aggregator.processPage(Aggregator.java:42)
        at org.elasticsearch.compute.operator.AggregationOperator.addInput(AggregationOperator.java:79)
        at org.elasticsearch.compute.operator.Driver.runSingleLoopIteration(Driver.java:214)
        at org.elasticsearch.compute.operator.Driver.run(Driver.java:139)
        at org.elasticsearch.compute.operator.Driver$1.doRun(Driver.java:327)
```

Also, not sure if this is supposed to work or not, or maybe return a more descriptive and user-friendly error message.

Reply:
I had a quick look at this and this seems to be the culprit.
This code works for me:

```
        protected PhysicalPlan rule(AggregateExec aggregate) {
            Set<FieldAttribute> fieldAttributes = new HashSet<>();

            PhysicalPlan plan = aggregate.transformDown(UnaryExec.class, exec -> {
...
                            if (af.field() instanceof FieldAttribute fieldAttribute) {
                                // We need to both mark the field to load differently, and change the spatial function to know to use it
                                fieldAttributes.add(fieldAttribute);
                                changedAggregates = true;
...
                if (exec instanceof FieldExtractExec fieldExtractExec && fieldAttributes.size() > 0) {
                    // Tell the field extractor that it should extract the field from doc-values instead of source values
                    if (fieldAttributes.retainAll(fieldExtractExec.attributesToExtract())) {
                        exec = fieldExtractExec.withForStats(fieldAttributes);
                    }
                }
                return exec;
...
```

Reply:
This was a bug, and I changed the Holder to a Set, and wrote a test in `spatial.csv-spec` that tests this, as well as a PhysicalPlanner test to test the correct behaviour of the planner.